pkg/kubelet/kubelet.go

```diff
@@ -759,6 +765,61 @@ func (kl *Kubelet) makeEnvironmentVariables(ns string, container *api.Container)
 	return result, nil
 }
 
+// runtimeEnvVarValue determines the value that an env var should take when a container
+// is started.  If the value of the env var is the empty string, the source of the env var
+// is resolved, if one is specified.
+//
+// TODO: preliminary factoring; make better
+func (kl *Kubelet) runtimeEnvVarValue(envVar api.EnvVar, pod *api.Pod) (string, error) {
+	runtimeVal := envVar.Value
+	if runtimeVal != "" {
+		return runtimeVal, nil
+	}
+
+	switch {
+	case envVar.ObjectInfo != nil:
+		return kl.objectInfoRuntimeValue(envVar.ObjectInfo, pod)
+	case envVar.HostInfo != nil:
+		return kl.hostInfoRuntimeValue(envVar.HostInfo)
+	}
+
+	return runtimeVal, nil
+}
+
+func (kl *Kubelet) objectInfoRuntimeValue(objInfo *api.ObjectReference, pod *api.Pod) (string, error) {
+	// TODO: use objInfo Kind instead of "Pod"
+	internalFieldPath, _, err := api.Scheme.ConvertFieldLabel(objInfo.APIVersion, "Pod", objInfo.FieldPath, "")
+	if err != nil {
+		return "", err
+	}
+
+	runtimeVal := ""
+
+	// TODO: currently there is no generic utility for retrieving a field of
+	// an object using a fieldPath expression.  For now, support only the
+	// Name and Namespace fields.
+	switch internalFieldPath {
+	case "metadata.name":
```

Comment:
I am eager to have a more well defined selector language.  I am all for
specifications, but JSON Path doesn't seem to fit the bill as a spec (and
it has a bunch of stuff we may or may not want like wildcards, filters,
slices, scripts...)

Once we have this, I think we want to rapidly converge all sorts of
self-accessing things to it (e.g. the "metadata volume" PR).

Do we write our own spec that is a subset of jq and JSONPath?


On Wed, Apr 15, 2015 at 9:21 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#discussion_r28483110>
> :
>
> > +}
> > +
> > +func (kl *Kubelet) objectInfoRuntimeValue(objInfo *api.ObjectReference, pod *api.Pod) (string, error) {
> > +	// TODO: use objInfo Kind instead of "Pod"
> > +	internalFieldPath, _, err := api.Scheme.ConvertFieldLabel(objInfo.APIVersion, "Pod", objInfo.FieldPath, "")
> > +	if err != nil {
> > +		return "", err
> > +	}
> > +
> > +	runtimeVal := ""
> > +
> > +	// TODO: currently there is no generic utility for retrieving a field of
> > +	// an object using a fieldPath expression.  For now, support only the
> > +	// Name and Namespace fields.
> > +	switch internalFieldPath {
> > +	case "metadata.name":
>
> @bgrant0607 <https://github.com/bgrant0607> I have no objection to moving
> to a more well defined field selector language.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739/files#r28483110>
> .
>